module/zfs/arc.c

```diff
@@ -4148,15 +4267,20 @@ arc_evict_state(arc_state_t *state, arc_buf_contents_t type, uint64_t spa,
 				ARCSTAT_BUMP(arcstat_evict_not_enough);
 			}
 
-			break;
+			if (!use_evcttq || sublists_left == 0)
```

Comment:
I think the logic is still not perfect here.  Since in `use_evcttq` case `scan_evicted` is accounted only over the last `ntasks` sublists, it might be 0, while we were successful on some other sublists earlier.

PS: The number of `use_evcttq` special cases continues to grow, which makes me wonder at which point two different functions or code paths would be cleaner.

Reply:
Lets say we have 8 sublists and 4 tasks, while only sublist 1 has something to evict. On the first iteration we run through sublists 0-3 and evict something, just not enough. On the second iteration we try sublists 4-7 and evict nothing.  At that point both `scan_evicted` and `sublists_left` will be 0 and we'll exit instead of evicting from sublist 1 again.

Reply:
I wonder if we should zero `scan_evicted` only when `sublists_left` goes to zero.  BTW, I think the current logic is broken in one more way: when `sublists_left` reaches zero, but there is more to evict, it never gets reset.

Reply:
Thanks for the hypothetical scenario. I thought hard, and removed the `use_evcttq` check.

Only just noticed the `sublist_left` comment, I'll take a look at that tomorrow.

Reply:
Oh, no, I did nothing here. I must have just thought hard, and then forgotten everything. Rechecking now.

Reply:
Alright, last push tackles both, I think.

I've lifted `scan_evicted` outside the loop, and now it returns to its previous meaning: the total amount evicted over _all_ sublists. The block to test and eject if we evicted nothing after scanning them all is now testing `scan_evicted == 0 && sublists_left == 0` and always breaks. Then, at the bottom of the loop, if we've run out of sublists but haven't evicted enough yet, we reset `sublists_left`, `sublist_idx` and `scan_evicted` and start fresh.